src/devices/SoftPwm/SoftwarePwmChannel.cs

```diff
@@ -105,106 +110,89 @@ public SoftwarePwmChannel(int pinNumber, int frequency = 400, double dutyCycle =
-                        _isStopped = false;
-                    }
-
-                    // Use the wait helper method to wait for the length of the pulse
-                    if (_usePrecisionTimer)
-                    {
-                        Wait(_pulseWidthInMilliseconds);
-                    }
-                    else
-                    {
-                        Task.Delay(TimeSpan.FromMilliseconds(_pulseWidthInMilliseconds)).Wait();
-                    }
-
-                    // The pulse if over and so set the pin to low and then wait until it's time for the next pulse
-                    _controller.Write(_pin, PinValue.Low);
+                    Thread.Yield();
+                    continue;
+                }
 
-                    if (_usePrecisionTimer)
-                    {
-                        Wait(_periodInMilliseconds - _pulseWidthInMilliseconds);
-                    }
-                    else
-                    {
-                        Task.Delay(TimeSpan.FromMilliseconds(_periodInMilliseconds - _pulseWidthInMilliseconds)).Wait();
-                    }
+                if (_pinHighTime != TimeSpan.Zero)
+                {
+                    _controller.Write(_pin, PinValue.High);
+                    Wait(_pinHighTime);
                 }
-                else
+
+                if (_pinLowTime != TimeSpan.Zero)
                 {
-                    if (!_isStopped)
-                    {
-                        _controller.Write(_pin, PinValue.Low);
-                        _isStopped = true;
-                    }
+                    _controller.Write(_pin, PinValue.Low);
+                    Wait(_pinLowTime);
                 }
             }
+
+            _controller.Write(_pin, PinValue.Low);
         }
 
         /// <summary>
         /// A synchronous wait is used to avoid yielding the thread
         /// This method calculates the number of CPU ticks will elapse in the specified time and spins
         /// in a loop until that threshold is hit. This allows for very precise timing.
         /// </summary>
-        /// <param name="milliseconds">The milliseconds to wait for</param>
-        private void Wait(double milliseconds)
+        /// <param name="timeout">The time to wait.</param>
+        private void Wait(TimeSpan timeout)
         {
-            long initialTick = _stopwatch.ElapsedTicks;
-            long initialElapsed = _stopwatch.ElapsedMilliseconds;
-            double desiredTicks = milliseconds / 1000.0 * Stopwatch.Frequency;
-            double finalTick = initialTick + desiredTicks;
-            while (_stopwatch.ElapsedTicks < finalTick)
+            if (_stopwatch is null)
+            {
+                Thread.Sleep(timeout);
+                return;
+            }
+
+            TimeSpan elapsed = _stopwatch.Elapsed;
+            TimeSpan target = elapsed + timeout;
+
+            while (target > _stopwatch.Elapsed)
```

Comment:
consider using DelayHelpers for that (we didn't have that when this was written I believe)

Reply:
No we did not :-) That was one of the very first PR done :-) Rest came later on

Reply:
Looks like it doesn't currently support taking a `TimeSpan`. That overload could be added and then everything could be done in terms of 100ns ticks (which is likely the maxmimum resolution any hardware can realistically support due to latency, etc).

Reply:
it would be useful to add such support, feel free to add it

Reply:
I'll have an update for this sometime later this afternoon

Reply:
Should be resolved. There is a new `Delay(TimeSpan time, bool allowThreadYield)` helper that allows delays as precise as 100ns (this will realistically be lower due to hardware latency in querying the timer).